File: tests/fixtures/vulnerable/injection.ts
// Vulnerable: SQL Injection (CG-001)
import { Pool } from 'pg';

const pool = new Pool();

async function getUser(userId: string) {
  // Direct string concatenation in SQL query
  const result = await pool.query("SELECT * FROM users WHERE id = '" + userId + "'");
  return result.rows[0];
}

async function searchUsers(name: string) {
  // Template literal in SQL query
  const result = await pool.query(`SELECT * FROM users WHERE name = '${name}'`);
  return result.rows;
}

// Vulnerable: Command Injection (CG-002)
import { exec, execSync } from 'child_process';

function runCommand(userInput: string) {
  exec("ls -la " + userInput);
}

function pingHost(host: string) {
  execSync(`ping -c 4 ${host}`);
}

// Vulnerable: Code Injection (CG-003)
function processExpression(expr: string) {
  return eval(expr);
}

function createHandler(code: string) {
  return new Function("data", code);
}

// Vulnerable: Insecure Regular Expression / ReDoS (CG-023)
function isValidEmail(input: string) {
  return new RegExp("^([a-zA-Z0-9]+)+@").test(input);
}

// Vulnerable: NoSQL Injection (CG-024) — whole request body as filter
async function login(users: any, req: any) {
  return users.findOne(req.body);
}

// Vulnerable: Open Redirect (CG-025) — redirect target from user input
function goNext(res: any, req: any) {
  res.redirect(req.query.next);
}

// Vulnerable: JWT Signature Bypass (CG-026) — accepts the "none" algorithm
function verifyToken(jwt: any, token: string, secret: string) {
  return jwt.verify(token, secret, { algorithms: ['none'] });
}

// Vulnerable: XML External Entity (CG-070) — entity substitution enabled
function parseXml(libxmljs: any, data: string) {
  return libxmljs.parseXml(data, { noent: true });
}

File: tests/fixtures/vulnerable/injection.py
# Vulnerable: SQL Injection (CG-001) - Python
import sqlite3

def get_user(user_id):
    conn = sqlite3.connect('app.db')
    cursor = conn.cursor()
    cursor.execute(f"SELECT * FROM users WHERE id = '{user_id}'")
    return cursor.fetchone()

# Vulnerable: Command Injection (CG-002) - Python
import os
import subprocess

def run_command(user_input):
    os.system("ls " + user_input)

def ping_host(host):
    subprocess.call(f"ping -c 4 {host}", shell=True)

# Vulnerable: Insecure Deserialization (CG-041) - Python
import pickle
import yaml

def load_data(data):
    return pickle.loads(data)

def load_config(config_str):
    return yaml.load(config_str)

# Vulnerable: Cross-Site Scripting (CG-010) - Python
from django.utils.safestring import mark_safe

def render_comment(comment):
    return mark_safe(comment)

# Vulnerable: Insecure Randomness (CG-022) - Python
import random

def generate_password_reset_token():
    return random.choice(range(1000000))

# Vulnerable: Insecure Regular Expression / ReDoS (CG-023) - Python
import re

def is_valid_email(value):
    return re.compile(r"^([a-zA-Z0-9]+)+@").match(value)

# Vulnerable: NoSQL Injection (CG-024) - Python
def login(users, request):
    return users.find_one(request.json)

# Vulnerable: Open Redirect (CG-025) - Python
def go_next(request):
    return redirect(request.args.get("next"))

# Vulnerable: JWT Signature Bypass (CG-026) - disables signature checking
def verify_token(jwt, token, key):
    return jwt.decode(token, key, options={"verify_signature": False})

# Vulnerable: XML External Entity (CG-070) - external entities resolved
def parse_xml(etree, data):
    parser = etree.XMLParser(resolve_entities=True)
    return etree.fromstring(data, parser)
